QuoridorEngine/src/core/QuoridorGame.cs

```diff
@@ -196,6 +196,109 @@ private void placeWall(QuoridorMove move)
             gameHistory.Add(move);
         }
 
+        /// <summary>
+        /// Check whether a path connecting current player's square
+        /// and his target baseline exists using DFS algorithm.
+        /// </summary>
+        /// <param name="isWhite">Current player is White</param>
+        /// <returns>True if path exists</returns>
+        private bool playerCanReachBaseline(bool isWhite)
+        {
+            // Pending squares to be explored
+            Stack<(int, int)> frontierSquares = new();
+
+            // Already visited squares
+            HashSet<(int, int)> visitedSquares = new();
+
+            // Get current player object
+            QuoridorPlayer currentPlayer = getTargetPlayer(isWhite);
+            Debug.Assert(currentPlayer != null);
+
+            // Temporarily save player's current row
+            int initialRow = currentPlayer.Row;
+
+            // Add player's current square in frontier as first square to be explored
+            frontierSquares.Push((currentPlayer.Row, currentPlayer.Column));
+            while (frontierSquares.Count != 0)
+            {
+                // Get a square (and remove it) from frontier
+                (int currentSquareRow, int currentSquareCol) = frontierSquares.Pop();
+
+                // Skip this square if it has already been visited
+                if (visitedSquares.Contains((currentSquareRow, currentSquareCol))) 
+                    continue;
+
+                // Store current square as visited
+                visitedSquares.Add((currentSquareRow, currentSquareCol));
+
+                // Temporarily update player's current row 
+                currentPlayer.Row = currentSquareRow;
+
+                // Check if player has reached goal
+                if (currentPlayer.IsInTargetBaseline())
+                {
+                    // Restore player's initial row
+                    currentPlayer.Row = initialRow;
+                    return true;
+                }
+
+                // Get current square's legal neighbours
+                List<(int, int)> legalNeighbours = getLegalNeighbourSquares(currentSquareRow, currentSquareCol);
+
+                // Sort neighbours by descending row if current
+                // player is black to reach his baseline faster 
+                if (!isWhite)
+                    legalNeighbours = legalNeighbours.OrderByDescending(x => x.Item1).ToList();
+
+                // Store current square's legal neighbours in frontier to be explored later
+                foreach ((int, int) neighbourSquare in legalNeighbours)
+                    frontierSquares.Push(neighbourSquare);
```

Comment:
Do not push already visited nodes into the stack to avoid unnecessary computations

P.S. I know this check is done above as well and the algorithm works as is, but this will help avoid removing already visited nodes from the stack over and over again for no reason